### PR TITLE
fix: webapp screen reader accessibility improvements

### PR DESCRIPTION
**Pull request by arjita-mitra**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
 - webapp screen reader accessibility improvements

- The **PR Description**
  - overall webapp screen reader accessibility improvements
----

# What's new in this PR?

### Issues

- https://docs.google.com/spreadsheets/d/1g5ENPrk7NivDOJy9vKoymXnDvu42G3rlR9oKzYslWBs/edit?pli=1#gid=0

### Solutions

- adding accessible controls for screen reader

#### How to Test

- need to run screen readers for areas mentioned in the doc
----
##### References
1. https://docs.google.com/spreadsheets/d/1g5ENPrk7NivDOJy9vKoymXnDvu42G3rlR9oKzYslWBs/edit?pli=1#gid=0
2. https://wearezeta.atlassian.net/browse/ACC-21

### CODE REVIEW

File: src/script/components/TextInput/TextInput.tsx [hunk 1]
```diff
@@ -111,13 +116,28 @@ const TextInput: React.ForwardRefRenderFunction<HTMLDivElement, UserInputProps>
...
       </label>
       {isFilled && !isSuccess && !isError && (
-        <button type="button" css={cancelButtonCSS} onClick={onCancel}>
+        <button
+          type="button"
+          css={cancelButtonCSS}
+          onClick={onCancel}
+          aria-label={t('accessibility.userProfileDeleteEntry')}
+          onKeyDown={event => {
+            if (event.shiftKey && event.key === 'Tab') {
```
Comment (przemvs) [inline src/script/components/TextInput/TextInput.tsx]: Please use isTabKey from KeyboardUtils :)

File: src/script/components/TextInput/TextInput.tsx [hunk 2]
```diff
@@ -111,13 +116,28 @@ const TextInput: React.ForwardRefRenderFunction<HTMLDivElement, UserInputProps>
...
+              // shift+tab from clear button should focus on the input field
+              setIsEditing(true);
+            } else if (event.key === 'Tab') {
```
Comment (przemvs) [inline src/script/components/TextInput/TextInput.tsx]: Please use isTabKey from KeyboardUtils :)

File: src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx [hunk 3]
```diff
@@ -182,20 +184,32 @@ const AccountInput: React.FC<AccountInputProps> = ({
...
+            setIsEditing(true);
+          }}
           onKeyDown={event => {
             if (event.key === 'Enter' && !event.shiftKey && !event.altKey) {
               event.preventDefault();
               onValueChange?.(input);
               (event.target as HTMLInputElement).blur();
+              // on enter save changes and close the editable field
+              setIsEditing(false);
+            } else if (event.key === 'Tab' && !!!input) {
```
Comment (przemvs) [inline src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx]: Please use isTabKey from KeyboardUtils :)

File: src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx [hunk 4]
```diff
@@ -182,20 +184,32 @@ const AccountInput: React.FC<AccountInputProps> = ({
...
+          inputWrapperRef={inputWrapperRef}
           onChange={({target}) => updateInput(target.value)}
-          onCancel={() => updateInput('')}
+          onCancel={() => {
+            updateInput('');
+            textInputRef.current?.focus();
+            setIsEditing(true);
+          }}
           onKeyDown={event => {
             if (event.key === 'Enter' && !event.shiftKey && !event.altKey) {
```
Comment (przemvs) [inline src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx]: I know is not your code, but please change event.key === 'Enter' to isEnterKey :)

**Review by przemvs — CHANGES_REQUESTED** [review]
few changes :)

**Review by V-Gira — COMMENTED** [review]
I don't understand why testing is failling on GroupCreationModal. It passes locally and CI passed with no issues last PR

**Comment by atomrc** [comment]
> I don't understand why testing is failling on GroupCreationModal. It passes locally and CI passed with no issues last PR

@V-Gira there is a typecheck error because the `TextInput` component has changed with required props, that are not passed in the `GroupCreationModal` component :)

File: src/script/components/TextInput/TextInput.tsx [hunk 5]
```diff
@@ -29,6 +29,7 @@ import {
   getInputCSS,
   getLabelCSS,
 } from 'Components/TextInput/TextInput.styles';
```
Comment (atomrc) [inline src/script/components/TextInput/TextInput.tsx]:
```suggestion
} from './TextInput.styles';
```

File: src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx [hunk 6]
```diff
@@ -99,7 +101,7 @@ const AccountInput: React.FC<AccountInputProps> = ({
     return () => {
       document.removeEventListener('mousedown', handleClickOutside);
     };
-  }, [textInputRef]);
+  }, [inputWrapperRef]);
```
Comment (atomrc) [inline src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx]: This will never trigger since it's a `ref`, react doesn't listen to `ref` changed

File: src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx [hunk 7]
```diff
@@ -182,20 +184,32 @@ const AccountInput: React.FC<AccountInputProps> = ({
           label={label}
           name={valueUie ? valueUie : fieldName}
           value={input}
+          inputWrapperRef={inputWrapperRef}
```
Comment (atomrc) [inline src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx:187]: This ref is never `set` (passed in a `ref={inputWrapperRef}` prop) it will always be `null` here

File: src/script/page/LeftSidebar/panels/Conversations/GroupedConversationsFolder.tsx [hunk 8]
```diff
@@ -55,7 +55,7 @@ const GroupedConversationsFolder: React.FC<GroupedConversationsFolderProps> = ({
   return (
     <li className="conversation-folder" data-uie-name="conversation-folder" data-uie-value={folder.name}>
       <GroupedConversationHeader onClick={() => toggle(folder.id)} conversationLabel={folder} isOpen={isExpanded} />
-      <ul css={css({listStyle: 'none', padding: 0})}>
+      <ul css={css({listStyle: 'none', padding: 0})} id={folder.id}>
```
Comment (atomrc) [inline src/script/page/LeftSidebar/panels/Conversations/GroupedConversationsFolder.tsx:58]: why is this needed?

File: src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx (hunk 6 above)
Comment (atomrc) [reply src/script/page/MainContent/panels/preferences/accountPreferences/AccountInput.tsx]: well, to be fair, the previous code should not work as well.. 🙃

File: src/script/page/LeftSidebar/panels/Conversations/GroupedConversationsFolder.tsx (hunk 8 above)
Comment (arjita-mitra) [reply src/script/page/LeftSidebar/panels/Conversations/GroupedConversationsFolder.tsx:58]: This is for aria-controls. 
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-controls
@atomrc

Comment (atomrc) [reply src/script/page/LeftSidebar/panels/Conversations/GroupedConversationsFolder.tsx:58]: thanks for the clarification 👍
